QUAL: Phpstan github action

frederic34 opened the pull request:
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/ecb5d387-b105-416b-9a97-f0e6c83463a2)

frederic34 committed:
try phpstan action

frederic34 committed:
try phpstan action

frederic34 commented:
@eldy the check is a little too long but errors are reported in comment

frederic34 committed:
try phpstan action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
try to cancel previous action

frederic34 committed:
try to cancel previous action

frederic34 committed:
add some ignore errors

eldy commented:
May be problem is because the phpstan is run on all the project. No way to run it only on modified files ?

mdeweerd commented:
You could try to enable the cache.

See https://staabm.github.io/2023/10/21/phpstan-result-cache-gotchas.html#github-actions for instance.


Also rename phpstan.neon to phpstan.neon.dist .

mdeweerd commented:
Possibly also reinstate:
```
        bootstrapFiles:
                - build/phpstan/bootstrap.php
```

Some extra constants need to be defined in there to allow the main.inc.php to complete.

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
try to use cache

frederic34 committed:
try to use cache

frederic34 committed:
try to use cache

frederic34 commented:
@mdeweerd thank's for the tips

frederic34 committed:
try to use cache

frederic34 committed:
try to use cache

frederic34 committed:
try to use cache

mdeweerd commented:
Currently your are testing the cache, so it's ok to use the PR index.

However, I'ld suggest that the final version restores the cache from the PR cache if it exists and otherwise from the develop branch.

Afterwards, if on the develop branch, store to the develop branches' cache, and otherwise to the PR cache in a pull-request.

mdeweerd commented:
Comme c'est vraiment très long, je suggère de limiter l'analyse à qqs fichiers seulement le temps de valider la mise en place du cache.

frederic34 committed:
try to use cache

frederic34 commented:
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/c1a7488a-97dc-4bc7-b3fa-890695e16982)

frederic34 commented:
@mdeweerd tu es au devcamps ?

mdeweerd commented:
Non, je suis juste un peu l'amélioriation de la qualité.

mdeweerd commented:
D'ailleurs j'ai un PR pour phan en attente: #26480 .

frederic34 commented:
Oui je l'ai vu

frederic34 committed:
test on htdocs only

frederic34 committed:
test on htdocs only

frederic34 committed:
test on htdocs only

frederic34 committed:
test on htdocs only

frederic34 commented:
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/7531c3c3-97d8-432b-85a5-6df2131fc123)

frederic34 committed:
test on htdocs only

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
tmp dir

tmp dir

tmp dir

tmp dir

tmp dir

frederic34 committed:
tmp dir

frederic34 committed:
fix some phpstan reports

frederic34 committed:
fix some phpstan reports

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix some phpstan reports

frederic34 commented:
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/1b606edb-5651-4ec0-a44e-a2ce5194df51)

frederic34 committed:
debug

frederic34 committed:
debug

frederic34 committed:
return to level 0 to debug cache

frederic34 committed:
return to level 0 to debug cache

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
clean

frederic34 committed:
clean

clean

clean

clean

clean

frederic34 committed:
clean

check with php version

check with php version

check with php version

check with php version

check with php version

check with php version

check with php version

check with php version

try to find error 139

frederic34 committed:
try to find error 139

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
try to find error 139

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
try to find error 139

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
remove duplicate

mdeweerd commented:
La mise en place du cache a bien avancé (récupération + sauvegarde), mais il y a toujours ce message de phpstan: "Result cache not used because the cache file does not exist." - donc il manque qqchse.

Perso l'expérience que j'ai c'est avec gitlab ou j'ai la tache suivante:

```yaml
phpstan:
  interruptible: true
  rules:
    - if: $CI_PIPELINE_SOURCE == "push" || $CI_PIPELINE_SOURCE == "trigger"
  stage: test
  image:
    name: ghcr.io/phpstan/phpstan
    entrypoint: [""]
  script:
    - phpstan analyse --no-progress --error-format gitlab | tee phpstan.json
    # - phpstan analyse -vvv
  cache:
    key: phpstan-$CI_COMMIT_REF_SLUG
    paths:
      - .tmp/
      - .cache/
      - vendor/
  artifacts:
    when: always
    reports:
      codequality: phpstan.json
```

Le cache de 'vendor' n'a pas d'utilité pour ce target de mon exemple, mais peut l'avoir quand on installe des modules pour analyse et évite de réinstaller à chaque fois.

Par contre je ne vois pas '.cache' dans la config pour dolibarr et aussi j'ai .tmp/ dans ma config, pas /tmp - le tmp n'est peut-être pas utile mais je n'ai pas cherché à optimiser ce point.

frederic34 commented:
le problème de cache vient de là
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/f523aed1-f8b5-4ad6-84e4-573059cd1bce)
j'ai eu des étapes où il le retrouvait

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
try to avoid error 139

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
try to exclude a file

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
phpstan doesnt like const VERSIONMIN=self::VERSIONMIN;

frederic34 commented:
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/91f3061c-8c1b-4552-878a-96762136883f)
when i get resultCache.php present, it say that cache was generated in the lest run

frederic34 commented on htdocs/core/db/DoliDB.class.php:
```diff
@@ -30,7 +30,7 @@
 abstract class DoliDB implements Database
 {
 	/** @var string Force subclass to implement VERSIONMIN */
-	const VERSIONMIN=self::VERSIONMIN;
```
@mdeweerd @eldy this make phpstan crash

mdeweerd replied:
I opened an [issue with phpstan](https://github.com/phpstan/phpstan/issues/10155) .

I guess that for know this line has to be commented so that phpstan can pass.  It kept phan happy and also helps enforce that the constant is defined in derived classes though.

mdeweerd replied:
I also found why the cache is not in '/tmp' in my project.

I have the [tmpDir option](https://phpstan.org/config-reference#caching) set:

```yaml
parameters:
  tmpDir: .cache/phpstan
```

If you rename phpstan.neon to phpstan.neon.dist, then you can set local parameters in phpstan.neon.  So for the github action you could create phpstan.neon with the specific location for the phpstan cache directory.
That helps avoid caching the entire /tmp directory which may not only hold phpstan cache.

frederic34 replied:
@mdeweerd phpstan bug has been fixed

frederic34 committed:
test

frederic34 commented:
seems the cache running, ~1 minute to run
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/3d12f24f-299a-48c6-b17d-5d9a0d596134)

frederic34 committed:
return to level 1

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 commented:
with level 9
![image](https://github.com/Dolibarr/dolibarr/assets/3624836/8d1f2f2b-e809-4983-8de0-909c7b2fb891)

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix special_code and rang missing

frederic34 committed:
fix travis

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
check phpstan bug with self constant reference

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix properties

frederic34 committed:
use separate phpstan neon

frederic34 committed:
use separate phpstan neon

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'origin/phpstan_action' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
test can fail

frederic34 committed:
test can fail

frederic34 committed:
test can fail

frederic34 committed:
test can fail

frederic34 committed:
test can fail

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'origin/phpstan_action' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 commented:
@eldy what's missing to merge ?

eldy commented on build/phpstan/bootstrap.php:
```diff
@@ -1,11 +1,13 @@
 if (! defined("NOHTTPSREDIRECT")) define("NOHTTPSREDIRECT", '1');
 global $conf, $langs, $user, $db;
-include_once __DIR__ . '/../../htdocs/main.inc.php';
+// include_once __DIR__ . '/../../htdocs/main.inc.php';
```
What is gain of removing call to main.inc.php ? This does not break knowledge of type of conf, user and db variable ?

frederic34 replied:
Test will never fail, until we remove `continue-on-error: true` from phpstan.yml but we can see errors

frederic34 committed:
boostrap

frederic34 committed:
Merge remote-tracking branch 'origin/phpstan_action' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
do not change original bootstrap

frederic34 committed:
fix

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
remove bad fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
remove fix

frederic34 committed:
Merge remote-tracking branch 'origin/phpstan_action' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'origin/phpstan_action' into phpstan_action

frederic34 committed:
add exclude

frederic34 committed:
do not hide problems

frederic34 committed:
remove ignore rule

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
test extension php

test extension php

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
fix

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'origin/phpstan_action' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge branch 'develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'upstream/develop' into phpstan_action

frederic34 committed:
Merge remote-tracking branch 'origin/phpstan_action' into phpstan_action
